### PR TITLE
unit-tests target: Fix backwards dependencies.

### PR DESCRIPTION
**Pull request by magnumripper**
...and refactor targets name from unit_tests to unit-tests.

### CODE REVIEW

**Comment by magnumripper** [comment]
This also obsoletes the "add to CI", because it will always happen.

**Comment by magnumripper** [comment]
Oh, and this immediately found a bug not really related to my changes here: There's some problem on Windows, possibly an `@EXE_EXT` thing. I never understood that crap, how is it needed at all?!

@jfoug feel free to push a fix to this branch

**Comment by jfoug** [comment]
Are you SURE you want this run on EVERY build???   If so, this really cuts down on what can be added to the unit test.     I would strongly recommend not doing this.    The name change re-factor is trivial.  The 'fixing' the mingw crap should be easy.  Fixing the ridiculous warning/error against perfectly working code is easy also.   But a FORCED run each time a 'default' build is done, .... hmmm.

**Comment by jfoug** [comment]
Ahh....    circle-ci will NOT succeed until I get the misc.c crap checked in.  This change will CORE any 32 bit build:

```diff
$ git diff misc.c
diff --git a/src/misc.c b/src/misc.c
index d06eaada9..ebc420e7b 100644
--- a/src/misc.c
+++ b/src/misc.c
@@ -257,7 +257,8 @@ void *strncpy_pad(void *dst, const void *src, size_t size, uint8_t pad)
        uint8_t *d = dst;
        const uint8_t *s = src;

-       if ((int64_t)size < 1)
+       // logically the same as:  if (size < 1) IF size were signed.
+       if (!size || size > (((size_t)-1) >> 1))
                return dst;

        while (*s && size) {
```

The reason is due to the typecast of (int64_t)size     (or any 64 bit typecast) on a 32 byte system will convert -1 into 0x00000000FFFFFFFF   and not -1

**Comment by jfoug** [comment]
This PR is dependent on #3587 to pass the CI on circle (the 32 bit stuff).

File: src/Makefile.in [hunk 1]
```diff
@@ -184,7 +184,7 @@ endif
 #  also need to figure out -mpreferred-stack-boundary=4 and add (or whatever the stack size is)
 #
 #########################################################
-default:
+default: unit-tests
 	$(MAKE) find_version
```
Comment (jfoug) [inline src/Makefile.in]: I am going to fail this, which you can override, due to the unit test being done (AND RUN) at each make.  Doing the build would be acceptable, if we created a 'separate' target for unit-tests-run:   or something like that.

ALSO, it should not be default: unit-tests, but
```
default:
	$(MAKE) unit-tests
	$(MAKE) find_version
	...
```

I 'think'

**Comment by magnumripper** [comment]
> Are you SURE you want this run on EVERY build??? If so, this really cuts down on what can be added to the unit test. I would strongly recommend not doing this. 

No I'm not. Or, we could have it on every build until/if it gets too heavy. Right now it adds about 1% of build time. That's a cheap insurance.

**Comment by jfoug** [comment]
>     Are you SURE you want this run on EVERY build??? If so, this
>     really cuts down on what can be added to the unit test. I would
>     strongly recommend not doing this.
>
> No I'm not. Or, we could have it on every build until/if it gets too 
> heavy. Right now it adds about 1% of build time. That's a cheap insurance.
>
Agreed (for now).  It is very light ATM.    I was just wondering if 
users might start to complain about these 'tests' that are done at every 
build.


Yes, for some systems (such as CPAN), and test IS required, but only 
prior to the install.  We do not (AFAIK) have an install target, so what 
we have done is to put the test into EVERY build. You go from making a 
tiny change due to a syntax error, do a make, and now something which 
used to take 5s now takes 30+s.  That is what I was stating.   Well, I 
am not sure about this, since the target is now based upon .o files.  It 
may only test IF something within the unit-tests set of data changes.  
So first build gets a unit-test.  Then a syntax error change in 
some_format_fmt_plug.c will only generate a normal build.

This might not be bad at all.   BUT I was also looking at unit testing 
things which are HARDER to test (such as compiler.c). There, you almost 
have to treat it like the current jtrts stuff, and build scripts to test.

**Comment by jfoug** [comment]
> Note that the build does NOT fail. The execution fails, with `make: ../run/unit_tests: Command not found`

It fails on build now, due to your force merge undoing the commit I made.  Yes, there were 2 things.

1.  there was a ridiculous warning (upforced to error), about unused return of fgets. WTF cares about it's return, if you know how to use feof().   BUT I did make a substantial, and beneficial change to unit-tests.c which your last force push has removed, where I stopped using fgets altogether, and now use fgetll().  That allowed much smaller buffers also, and no worry about unowned input files, having super long lines.

2. there was a problem during runtime for 32 bit builds, in the strnzcpy_pad function.  This 'has' been updated, and should be ok.

**Comment by magnumripper** [comment]
> It fails on build now, due to your force merge undoing the commit I made.

Ouch, sorry I missed that. I'm used to getting a mail for every push to the repo but GiHub dropped that.

**Comment by magnumripper** [comment]
I found your commit and pushed it over here. However, as I stated I never indended to rename any **file**, only the make **target**. We'll see how this goes.

**Comment by magnumripper** [comment]
So we're back from where we started: Only Circle fails, it builds fine but then can't find the executable.
```
Configure finished.  Now "make -s clean && make -sj4" to compile.
Now Running the Unit Tests
../run/unit-tests.exe: ../run/unit-tests.exe: cannot execute binary file
make: *** [Makefile:1667: unit-tests] Error 126
Exited with code 2
```

**Comment by jfoug** [comment]
> as I stated I never indended to rename any **file**, only the make **target**. Anyway we'll see how this goes.

I had not read the comment you made on that prior commit.  I simply saw this PR, saw it was failing.  Followed the trail, fixed the fgets() warnings.  Saw you made some name changes, so change the rest.  Only later did I see the comment you made about the source being named properly.

Then when that build failed, I followed the trail again, and saw it was 'during' the run, and that the strnzcpy_pad was not updated, causing the 32 bit builds to core at runtime.

I see circle is still down, I will dig into this.

**Comment by jfoug** [comment]
Does mingw need this:

$(CC) -o ../run/unit-tests@EXE_EXT@ $(LDFLAGS) -D_MISC_C_UNIT_TEST -D_JOHN_MISC_NO_LOG $(UNIT_TEST_OBJS)


???   I will give that a try, since it is such a trivial change to make, and should not impact anything where that macro would NOT be set (i.e. almost everything other than mingw, I think).

**Comment by jfoug** [comment]
I added the @exe_ext@ to 2 locations.

1. to the default: requirement
2. to the target for unit-tests

Hopefully that is the only places it should be.

NOTE, I am not sure where the 'empty' merge commit came from.

**Comment by jfoug** [comment]
Ok, I am about to give up.  

Can we get circle CI to build WITHOUT make -s  so that I can see just WTF it is doing?

Ok, I hope I got the Circle code to now build without the -s.    Once we know wtf is going on, we can roll-back that commit.

**Comment by jfoug** [comment]
Ok, I give up.  I will have to make a few cosmetic changes, once we FIGURE this crap out.

Is there any way to log into an already built CircleCI environment, like a shell access?  We really need to see just what the problem is.  Using CicleCI and throwing different hypothetical Makefiles at it hoping one will stick is not the way.

**Comment by jfoug** [comment]
I have this change, but simply hate throwing things at this problem (simply spawning off builds).

```diff
$ git diff
diff --git a/src/Makefile.in b/src/Makefile.in
index cc3701bfa..f096163dd 100644
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -184,7 +184,7 @@ endif
 #  also need to figure out -mpreferred-stack-boundary=4 and add (or whatever the stack size is)
 #
 #########################################################
-default:       ../run/unit-tests@EXE_EXT@
+default:       unit-tests
        $(MAKE) find_version
        @$(MAKE) $(PROJ) \
                JOHN_OBJS="$(JOHN_OBJS) @CC_ASM_OBJS@" \
@@ -799,12 +799,14 @@ tests/memory.o:   memory.c arch.h misc.h jumbo.h autoconfig.h memory.h common.h jo
 # keep the 'easy name' build target of unit-tests   The 'real' target is ../run/unit-tests[.exe]
 unit-tests:    ../run/unit-tests@EXE_EXT@

+# here is link line from john
+#      $(LD) $(JOHN_OBJS) $(LDFLAGS) @OPENSSL_LIBS@ @COMMONCRYPTO_LIBS@ @OPENMP_CFLAGS@ @GMP_LIBS@ @SKEY_LIBS@ @REXGEN_LIBS@ @CL_LIBS@ @PTHREAD_CFLAGS@ @PTHREAD_LIBS@ @M_LIBS@ @Z_LIBS@ @DL_LIBS@ @CRYPT_LIBS@ @BZ2_LIBS@ @ZTEX_LIBS@ aes/aes.a secp256k1/secp256k1.a ed25519-donna/ed25519-donna.a -o $@
+
 ../run/unit-tests@EXE_EXT@:    $(UNIT_TEST_OBJS)
-       $(CC) $(LDFLAGS) -D_MISC_C_UNIT_TEST -D_JOHN_MISC_NO_LOG $(UNIT_TEST_OBJS) -o $@
+       $(LD) $(UNIT_TEST_OBJS) $(LDFLAGS) @OPENSSL_LIBS@ @COMMONCRYPTO_LIBS@ -o $@
        @ echo "Now Running the Unit Tests"
-       @ ls -l ../run
-       @ ../run/unit-tests
-       @ ../run/unit-tests@EXE_EXT@
+       @ # ls -l ../run
+       ../run/unit-tests@EXE_EXT@
        @ # Extra testing past the unit-tests code commented out for now until I figure out how best to do this.
        @ # echo "Now Testing john 'external' mode"
        @ # ../run/john -conf=tests/john-tc.conf -stdout -extern=double 2> /dev/null > tests/external.tst

```

This uses $(LN) and not $(CC)   Also, a couple libs are linked in (but I do not think they are required).  Also, I removed the @ from the execution line.  Hopefully, if there is some error message about cant find lib of something like that, it may show up.

But again, without being able to test this, I HATE simply tossing it against the wall.

**Comment by jfoug** [comment]
Ok, found out how to SSH into a circle instance (I am there debugging now)   https://circleci.com/docs/2.0/ssh-access-jobs/

Here is what I see:

bash: ../run/unit-tests.exe: cannot execute binary file: Exec format error

But now, I can mess with Makefile, until I get this damn thing working!

**Comment by jfoug** [comment]
Ok, I completed the make. Now I get this:

../run/john.exe: cannot execute binary file: Exec format error

NOTE, this is building -mavx    I wonder if we are simply cross compiling and THIS will never run. This brings up a VERY interesting issue!!  This will not be able to be done as it is today.  It will not run properly on cross compilers!!

Ok, here is how mingw testing is done, with the mingw-circle.sh script:

```sh
build  flavor1
mv ../run/john.exe   ../run/john-flavor1.exe
build  flavor2
mv ../run/john.exe   ../run/john-flavor2.exe
build  flavor3
mv ../run/john.exe   ../run/john-flavor3.exe

# Then later on in the script, it tests IN a wine environment.
echo ""
echo ""
echo ""
echo ""
echo '******************************************************************************'
echo "now testing the 32-bit SSE2 Windows build"
echo '******************************************************************************'
echo ""
# crazy testing!
cd /base/JohnTheRipper/run
export WINEDEBUG=-all  # suppress wine warnings
/usr/bin/wine john.exe --list=build-info
echo "[Disabled:Formats]" > john-local.conf
/usr/bin/wine john.exe --test-full=0
```

So the ONLY way to get it working as we have it today, is to also run with WINE.  Let me see if I can get Makefile.in working to do this

**Comment by jfoug** [comment]
Nope, circle is now running 'fine'.  I will let this finish up, BUT again, we need to figure out how to deal with cross compiling!

**Comment by magnumripper** [comment]
Thanks, very interesting.

I have an idea... what if a `--test-full` run would start with these new tests instead of having it as a separate binary? Although the dependency on the separate test vector files isn't great then.

**Comment by magnumripper** [comment]
```diff
-default:       ../run/unit-tests@EXE_EXT@
+default:       unit-tests
```
This is one reason I did not rename the files, only the target. The first line is obviously wrong, it's supposed to specify the *make target* and not some exe file.

**Comment by jfoug** [comment]
Thanks, very interesting.
>
> I have an idea... what if a |--test-full| run would start with these 
> new tests instead of having it as a separate binary? Although the 
> dependency on the separate test vector files isn't great then.
>
There are pros/cons to both methods. I initially was very hesitant about 
putting in a pre-qualification test, which must be passed prior to 
running a make. However, after working with it, thinking about it, 
implementing it, I AM sold on the concept. HOWEVER, as we are seeing 
here (and would have almost NO way to work around for cross compiling), 
this layout can not handle all build environments.  We saw that for 
Windoze builds under Fedora on CircleCI.  Yes, we 'did' get that worked 
through, but ONLY on circle. I tests env var and ONLY perform that 
special stuff on a circle build, and only if the gcc is 1 of 2 flavors. 
   So we have dealt with that cross compiler, BY figuring it out, and 
then coding specifically for it.  But in general that is a 100% non-starter.

Now, if we can detect cross compiling (at make time), that may be 
another thing.  We may be able to run this pre-qualification stuff, IF 
we can detect that we are building a 'native' process, which can run.  
Otherwise we simply provide the target in a makefile for that 
environment, and do nothing.

one additional thing to look at.  IF we keep this pre-qualification 
method, should this logic not be in the configure script??  This train 
of thought may really derail the scope of this project, so if you 
believe so, then simply state this, and we will  never discuss it again ;)

**Comment by magnumripper** [comment]
Yes it did occur to me `configure` does know whether we're cross-compiling or not. So perhaps autoconf could add it to the normal build target, only when not cross compiling.

**Comment by jfoug** [comment]
That was kinda my idea.  Something like:

```diff
default:  @UNIT_TESTS_IF_NOT_XCOMIPLE@
```
In this way, the unit-tests target is in the makefile. But it we are cross compiling, it is not run (by default).

Note, the target doing an auto run (and failing), is 'ok' if the only goal is to build the exe, and then copy it to the proper machine to run.  So I do not see the need to also have a configure replace-macro for the unit-tests launcher line.

**Comment by jfoug** [comment]
I am going to make a test push (cross compiling).  It also has the 'special' logic for running under Win32 for CircleCI commented out, to make sure that it passes the Circle environment (i.e. the unit-tests will not be done, unless they can be run).  I think I have it done properly, but we will soon see, lol.

**Comment by jfoug** [comment]
So far, the new build process looks good. All 3 of the windows builds worked properly (i.e. did not do any unit-tests run).  Now, if the other builds DO perform the unit-tests on a preliminary fashion, prior to the default make, then I will approve the changes, and we can merge this beast.

**Comment by jfoug** [comment]
Ok, that build worked great, BUT there were no unit-tests on Circle at all.  I think I am going to touch the Makefile.legacy, so that we get some older 32 bit linux testing doing the unit-tests logic.

Once those changes pass the CI's, I will approve the change request I made, thus clearing up this PR.

**Pull request merged**